### PR TITLE
New features and fixes

### PR DESCRIPTION
**Pull request by neon12345**

### CODE REVIEW

File: src/EmuFs.cc [hunk 1]
```diff
@@ -72,6 +72,22 @@ void EmuFile::update(dev_t device, ino_t inode, uint64_t size) {
   size_ = size;
 }
 
+#ifndef MFD_CLOEXEC
+#define MFD_CLOEXEC 0x0001u
+#endif
+#ifndef MFD_ALLOW_SEALING
+#define MFD_ALLOW_SEALING 0x0002u
+#endif
```
Comment (rocallahan) [inline src/EmuFs.cc]: These should go in `kernel_supplement.h`

File: src/EmuFs.cc [hunk 2]
```diff
@@ -72,6 +72,22 @@ void EmuFile::update(dev_t device, ino_t inode, uint64_t size) {
...
+
+#ifndef __NR_memfd_create
+#define __NR_memfd_create 319
+#endif
```
Comment (rocallahan) [inline src/EmuFs.cc]: Just use `NativeArch::memfd_create`. This has the bonus feature of also working on x86-32.

File: src/EmuFs.cc [hunk 3]
```diff
@@ -72,6 +72,22 @@ void EmuFile::update(dev_t device, ino_t inode, uint64_t size) {
...
+
+static int memfd_create(const char *__name, unsigned int __flags)
+{
+  return syscall(__NR_memfd_create, __name, __flags);
+}
```
Comment (rocallahan) [inline src/EmuFs.cc]: Call this `rr_memfd_create` so it doesn't conflict with a libc version when that appears.

File: src/EmuFs.cc [hunk 4]
```diff
@@ -86,14 +102,21 @@ void EmuFile::update(dev_t device, ino_t inode, uint64_t size) {
        << "-inode-" << orig_inode << "-" << path_tag;
   string real_name = name.str().substr(0, 255);
 
-  ScopedFd fd =
-      open(real_name.c_str(), O_CREAT | O_EXCL | O_RDWR | O_CLOEXEC, 0600);
+  int res = memfd_create(real_name.c_str(), MFD_CLOEXEC);
```
Comment (rocallahan) [inline src/EmuFs.cc]: Can you confirm that you really need this in practice? I'd prefer to not add another testing path without a confirmed need.

File: src/replay_syscall.cc [hunk 5]
```diff
@@ -673,8 +673,17 @@ static void finish_shared_mmap(ReplayTask* t, AutoRemoteSyscalls& remote,
   // For full generality, we also need to emulate direct file
   // modifications through write/splice/etc.
   off64_t offset_bytes = page_size() * offset_pages;
-  if (ssize_t(buf.data.size()) !=
-      pwrite64(emufile->fd(), buf.data.data(), buf.data.size(), offset_bytes)) {
+  ssize_t res = 0;
+  do
+  {
```
Comment (rocallahan) [inline src/replay_syscall.cc]: `do {`

File: src/replay_syscall.cc [hunk 6]
```diff
@@ -673,8 +673,17 @@ static void finish_shared_mmap(ReplayTask* t, AutoRemoteSyscalls& remote,
...
+    ssize_t c = pwrite64(emufile->fd(), buf.data.data() + res, buf.data.size() - res, offset_bytes + res);
+    if (c == -1)
+    break;
```
Comment (rocallahan) [inline src/replay_syscall.cc]:
````
    if (c == -1) {
      break;
    }
````

Comment (rocallahan) [inline src/replay_syscall.cc]: If `pwrite64` returns 0 then I guess we need to abort because the device is full...

File: src/record_syscall.cc [hunk 7]
```diff
@@ -2480,7 +2480,8 @@ static Switchable rec_prepare_syscall_arch(RecordTask* t,
         case Arch::SETOWN:
         case Arch::SETOWN_EX:
         case Arch::GETSIG:
-        case Arch::SETSIG:
+        case Arch::SETPIPE_SZ:
+        case Arch::GETPIPE_SZ:
```
Comment (rocallahan) [inline src/record_syscall.cc]: Why did you remove `SETSIG`?

File: src/test/fcntl_pipe_sz.c [hunk 8]
```diff
@@ -0,0 +1,33 @@
+/* -*- Mode: C; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include "rrutil.h"
+
+int main(void) {
+
```
Comment (rocallahan) [inline src/test/fcntl_pipe_sz.c]: Remove blank line

File: src/test/fcntl_pipe_sz.c [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  int p[2];
+  int size;
+  int nsize;
+
+  test_assert(pipe(p) != -1);
+
+  size = fcntl(p[0], F_GETPIPE_SZ);
+
+  if (size == -1 && errno == EINVAL)
+  {
```
Comment (rocallahan) [inline src/test/fcntl_pipe_sz.c]: `{` on previous line

File: src/test/fcntl_pipe_sz.c [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+    // should we succeed when PIPE_SZ is not supported?
```
Comment (rocallahan) [inline src/test/fcntl_pipe_sz.c]: No. It's supported since 2.6.35 and rr definitely doesn't work at all before 3.6 or thereabouts, so these fcntls will definitely be supported.

**Comment by rocallahan** [comment]
Thanks. These basically look great.

File: src/record_syscall.cc (hunk 7 above)
Comment (neon12345) [inline src/record_syscall.cc]: bug

File: src/replay_syscall.cc (hunk 6 above)
Comment (neon12345) [inline src/replay_syscall.cc]: According to the man page EOF is for read ony. I would expect write to fail with -1 if the disc is full.

File: src/EmuFs.cc (hunk 4 above)
Comment (neon12345) [inline src/EmuFs.cc]: The problem is that /dev/shm is fixed in size to e.g. half the ram and I don't want to tweak this if needed when there is a better interface. I guess it also has the benefit that there is no tmpfs overhead.

Comment (rocallahan) [inline src/EmuFs.cc]: OK then...

File: src/replay_syscall.cc (hunk 6 above)
Comment (rocallahan) [inline src/replay_syscall.cc]: I guess that's true but I think we should abort for 0 instead of retrying, anyway. It's no more complicated to do that.
